### PR TITLE
Support the ESPlayer for playing element stream on TV

### PR DESCRIPTION
**Pull request by yourina**
### Description of Change ###
Add the ESPlayer

### Bugs Fixed ###
N/A

### API Changes ###
Added: TVESPlayer class

### Behavioral Changes ###
N/A

### CODE REVIEW

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        public ESMatroskaColor? matroskaColorInfo;
+    }
+
+
+    /// <summary>
+    /// Represents stream information for video stream which is demuxed from application,
+    /// contains codec data, mime type, width, height, max width, max hgeight and frame
+    /// rate(num, den).
+    /// </summary>
+    public class VideoStreamInfo
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:103]:
```suggestion
    public struct VideoStreamInfo

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        /// <summary>
+        /// Codec data for the associated video stream
+        /// </summary>
+        public byte[] codecData { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:108]:
```suggestion
        public byte[] CodecData { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Mime type for the associated video stream
+        /// </summary>
+        public VideoMimeType mimeType { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:112]:
```suggestion
        public VideoMimeType MimeType { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// width for the associated video stream
+        /// </summary>
+        public int width { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:116]:
```suggestion
        public int Width { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Height for the associated video stream
+        /// </summary>
+        public int height { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:120]:
```suggestion
        public int Height { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Max width for the associated video stream
+        /// </summary>
+        public int maxWidth { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:124]: What is max width?
```suggestion
        public int MaxWidth { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Max height for the associated video stream
+        /// </summary>
+        public int maxHeight { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:128]:
```suggestion
        public int MaxHeight { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Numerator of framerate for the associated video stream
+        /// </summary>
+        public int num { get; set; }
+        /// <summary>
+        /// Denominator of framerate for the associated video stream
+        /// </summary>
+        public int den { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:136]: Why not use usually used formatting like a `60`, `24`, `120`?

``` c#
public struct ESFrameRate 
{
    public int Numerator {get;set;}
    public int Denominator {get;set;}
}
```
```suggestion
        public ESFrameRate FrameRate { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public class AudioStreamInfo
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs:139]:
```suggestion
    public struct AudioStreamInfo

```

File: src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+            set
+            {
+                _esImpl.VideoStreamInfo = value;
+            }
+        }
+
+        /// <summary>
+        /// Get the time for playing of player
+        /// </summary>
+        public TimeSpan PlayingTime
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs:77]: What is diffrent from `Position` on MediaPlayer

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        public StreamEventArgs(StreamType type)
+        {
+            Type = type;
+        }
+
+        /// <summary>
+        /// The type of stream for ESPlayer
+        /// </summary>
+        public StreamType Type { get; set; }
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs:54]:
```suggestion
        public StreamType StreamType { get; set; }

```

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        /// <summary>
+        /// The buffer status of ESPlayer
+        /// </summary>
+        public BufferStatus BufferStatus { get; internal set; }
+    }
+
+    /// <summary>
+    /// Arguments for the event that is raised when error is cccurred
+    /// </summary>
+    public class ErrorEventArgs : EventArgs
```
Comment (myroot) [inline src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs:28]:
```suggestion
    public class ErrorOccurredEventArgs : EventArgs

```

**Review by myroot — CHANGES_REQUESTED** [review]

**Comment by rookiejava** [comment]
Thanks for contribution! Two simple questions after a quick glance review.

1. who is main target for this API? Can 3rd party developers can use ESPlayer? Would great if you can provide sample as well.

2. is there special reason why don't you add classes into existing name space?

File: src/Tizen.TV.UIControls.Forms/MediaPlayer.cs [hunk 13]
```diff
@@ -92,7 +92,7 @@ public class MediaPlayer : Element
         /// </summary>
         public static readonly BindableProperty IsBufferingProperty = IsBufferingPropertyKey.BindableProperty;
 
-        IPlatformMediaPlayer _impl;
+        protected IPlatformMediaPlayer _impl;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/MediaPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/Renderer/MediaPlayerImpl.cs [hunk 14]
```diff
@@ -30,7 +30,7 @@ namespace Tizen.TV.UIControls.Forms.Renderer
 {
     public class MediaPlayerImpl : IPlatformMediaPlayer
     {
-        Player _player;
+        protected Player _player;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/Renderer/MediaPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        ITVESPlayer _esImpl;
+
+        /// <summary>
+        /// Initializes a new instance of the TVESPlayer class.
+        /// </summary>
+        public TVESPlayer() : base()
+        {
+            if (_impl == null) {
+                _impl = CreateMediaPlayerImpl();
+            }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+            _esImpl.BufferStatusChanged += SendBufferStatusChanged;
+            _esImpl.EOSEmitted += SendEOSEmitted;
+            _esImpl.ErrorOccurred += SendErrorOccurred;
+            _esImpl.StreamReady += SendStreamReady;
+            _esImpl.SeekReady += SendSeekReady;
+        }
+
+        protected override IPlatformMediaPlayer CreateMediaPlayerImpl()
+        {
+            return DependencyService.Get<ITVESPlayer>(fetchTarget: DependencyFetchTarget.NewInstance) as IPlatformMediaPlayer;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 17]
```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using Tizen.TV.UIControls.Forms;
+using Tizen.TV.Multimedia;
+
+namespace Tizen.TV.Extension.UIControls.Forms
+{
+    public interface ITVESPlayer : IPlatformMediaPlayer
+    {
+        VideoStreamInfo VideoStreamInfo { set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: I will change this property to method. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+    /// <summary>
+    ///  Represents one of es packet which is demuxed from application, contains stream type, es packet buffer, pts and duration.
+    /// </summary>
+    public class ESPacket
+    {
+        /// <summary>
+        /// Stream type of ESPacket
+        /// </summary>
+        public StreamType type { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: I will fixed it. Thank you.

[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Buffer containing demuxed es packet
+        /// </summary>
+        public byte[] buffer { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Sorry for bothering you to review this coding convention(or rule). Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// PTS(Presentation Time Stamp) of es packet
+        /// </summary>
+        public ulong pts { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Duration of es packet
+        /// </summary>
+        public ulong duration { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        /// The matroska color information. This value is only for video packet. If you set this value on a packet of other type, you can see an error when you submit the packet.
+        /// </summary>
+        public ESMatroskaColor? matroskaColorInfo { get; set; }
+    }
+
+    /// <summary>
+    /// Represents one of es packet which includes handle of data inside the trust zone,
+    /// contains stream type, handle for encrypted es data packet, handle size, pts and duration.
+    /// </summary>
+    public struct ESHandlePacket
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        /// <summary>
+        /// Stream type of ESPacket
+        /// </summary>
+        public StreamType type;
+        /// <summary>
+        /// Handle for encrypted es packet inside the trust zone
+        /// </summary>
+        public uint handle;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: The value of this handle is obtained from APIs of `Tizen.TV.Security`. APIs of `Tizen.TV.Security` use uint type for handle. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+    /// <summary>
+    /// Represents one of es packet which includes handle of data inside the trust zone,
+    /// contains stream type, handle for encrypted es data packet, handle size, pts and duration.
+    /// </summary>
+    public struct ESHandlePacket
+    {
+        /// <summary>
+        /// Stream type of ESPacket
+        /// </summary>
+        public StreamType type;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+        /// </summary>
+        public int maxWidth { get; set; }
+        /// <summary>
+        /// Max height for the associated video stream
+        /// </summary>
+        public int maxHeight { get; set; }
+        /// <summary>
+        /// Numerator of framerate for the associated video stream
+        /// </summary>
+        public int num { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 26]
```diff
@@ -0,0 +1,395 @@
...
+            {
+                case ESPlayerState.Playing:
+                case ESPlayerState.Paused:
+                    esPlayer.Stop();
+                    esPlayer.Close();
+                    break;
+            }
+        }
+
+        async void OnReadyToPrepare(Tizen.TV.Multimedia.StreamType streamType)
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 27]
```diff
@@ -0,0 +1,395 @@
...
+        }
+
+        public SubmitStatus SubmitPacket(ESHandlePacket packet)
+        {
+            return esPlayer.SubmitPacket(packet.ToESHandlePacket()).ToSubmitStatus();
+        }
+
+        public async Task<int> Seek(int ms)
+        {
+            TimeSpan time = new TimeSpan(0, 0, 0, 0, ms);
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 28]
```diff
@@ -0,0 +1,395 @@
...
+                case Tizen.TV.Multimedia.StreamType.Audio:
+                    StreamReady?.Invoke(this, new StreamEventArgs(StreamType.Audio));
+                    break;
+                case Tizen.TV.Multimedia.StreamType.Video:
+                    StreamReady?.Invoke(this, new StreamEventArgs(StreamType.Video));
+                    break;
+            }
+        }
+
+        async void OnSeekReady(TM.StreamType streamType, TimeSpan time)
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 29]
```diff
@@ -0,0 +1,395 @@
...
+        {
+            ErrorOccurred?.Invoke(sender, new ErrorEventArgs(e.ErrorType.ToErrorType()));
+        }
+
+        void OnEOSEmitted(object sender, EOSEventArgs e)
+        {
+            EOSEmitted?.Invoke(sender, EventArgs.Empty);
+        }
+
+        async Task ChangeToIdleState()
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 30]
```diff
@@ -0,0 +1,395 @@
...
+        {
+            switch (esPlayer.GetState())
+            {
+                case ESPlayerState.Playing:
+                case ESPlayerState.Paused:
+                    esPlayer.Stop();
+                    esPlayer.Close();
+                    break;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 31]
```diff
@@ -0,0 +1,395 @@
...
+        public event EventHandler<ErrorEventArgs> ErrorOccurred;
+        public event EventHandler<StreamEventArgs> StreamReady;
+        public event EventHandler<SeekEventArgs> SeekReady;
+        public event EventHandler<BufferStatusEventArgs> BufferStatusChanged;
+
+        public bool UsesEmbeddingControls { get; set; }
+
+        public bool AutoPlay { get; set; }
+
+        public bool AutoStop { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/Renderer/TVESPlayerImpl.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 1 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 2 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 3 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 4 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 5 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 6 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 7 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 8 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: This is good idea. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs (hunk 9 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs (hunk 10 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs]: Actually, `PlayingTime` is same with `Position` except of `Type`

File: src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+        /// </summary>
+        public SubmitStatus SubmitPacket(ESHandlePacket packet)
+        {
+            return _esImpl.SubmitPacket(packet);
+        }
+
+        /// <summary>
+        /// Resumes the es contents playback.
+        /// </summary>
+        public void Resume()
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs (hunk 11 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs [hunk 33]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        /// <summary>
+        /// The time for seeking of TVESPlayer
+        /// </summary>
+        public TimeSpan Time { get; set; }
+        /// <summary>
+        /// The type of stream for TVESPlayer
+        /// </summary>
+        public StreamType Type { get; set; }
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+        /// <summary>
+        /// The type of stream for ESPlayer
+        /// </summary>
+        public StreamType Type { get; set; }
+    }
+
+    /// <summary>
+    /// Arguments for the event that is raised when the TVESPlayer is ready to accept the first es packet of the point to be moved for each stream during seeking player.
+    /// </summary>
+    public class SeekEventArgs : EventArgs
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs (hunk 12 above)
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs]: Fixed it. Thank you.

File: src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs [hunk 35]
```diff
@@ -0,0 +1,77 @@
...
+        /// The type of error from ESPlayer
+        /// </summary>
+        public ErrorType ErrorType { get; internal set; }
+    }
+
+
+    /// <summary>
+    /// Arguments for providing stream infomation 
+    /// </summary>
+    public class StreamEventArgs : EventArgs
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.Extension.UIControls.Forms/TVESEventArgs.cs]: Fixed it. Thank you.

**Comment by yourina** [comment]
> 
> 
> Thanks for contribution! Two simple questions after a quick glance review.
> 
>     1. who is main target for this API? Can 3rd party developers use ESPlayer? Would great if you can provide sample as well.
> 
>     2. is there special reason why don't you add classes into existing name space?

Thank you for your question. I had like to reply your question. 
About first question, this API will be used only by 2nd party developers unfortunately because APIs of `ESPlayer` is only supported for 2nd party not yet. But, I hope that 3rd party devloper can use these APIs someday. 
And I think that these API should be added into other namespaces because these APIs are more dependent on TV product. But, it is better to support these APIs in `Tizen.TV.UIControls.Forms` namespace. If the purpose of other APIs in `Tizen.TV.UIControls.Forms` is providing the operation on TV product.  
Thank you.

**Comment by yourina** [comment]
> 
> 
> Thanks for contribution! Two simple questions after a quick glance review.
> 
>     1. who is main target for this API? Can 3rd party developers use ESPlayer? Would great if you can provide sample as well.
> 
>     2. is there special reason why don't you add classes into existing name space?

File: src/Tizen.TV.UIControls.Forms/Renderer/TVESPlayerImpl.cs [hunk 36]
```diff
@@ -0,0 +1,420 @@
...
+                        if (parentArea.Width == 0 || parentArea.Height == 0)
+                        {
+                            await Task.Delay(1);
+                            parentArea = renderer.NativeView.Geometry;
+                        }
+                        bound = parentArea;
+                    }
+                    var roiBound = bound.ToMultimedia();
+                    esPlayer.SetDisplayRoi(roiBound.X, roiBound.Y, roiBound.Width, roiBound.Height);
+                }
```
Comment (myroot) [inline src/Tizen.TV.UIControls.Forms/Renderer/TVESPlayerImpl.cs:412]: Need to refactoring, ESPlayer is different from Tizen.Player on `DisplayMode`, 

Tizen.Player explicitly has ROI mode, in this mode video was displayed on roi area and if not ROI mode, video displayed on window area, 
But, ESPlayer has no ROI mode, So, it seem to no way to return full screen after call `SetDisplayRoi`

File: src/Tizen.TV.UIControls.Forms/TVESPlayer.cs [hunk 37]
```diff
@@ -0,0 +1,155 @@
...
+    public class TVESPlayer : MediaPlayer
+    {
+        ITVESPlayer _esImpl;
+
+        /// <summary>
+        /// Initializes a new instance of the TVESPlayer class.
+        /// </summary>
+        public TVESPlayer() : base()
+        {
+        }
```
Comment (myroot) [inline src/Tizen.TV.UIControls.Forms/TVESPlayer.cs:34]: Remove empty constructor

File: src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs [hunk 38]
```diff
@@ -0,0 +1,294 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Tizen.TV.UIControls.Forms
+{
+    /// <summary>
+    /// Enumerator for stream type of es stream
+    /// </summary>
+    public enum StreamType
```
Comment (myroot) [inline src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs:22]: It seems need a sub namespace or rename, 
StreamType is too common to use
```suggestion
    public enum ESPlayerStreamType
```

File: src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs [hunk 39]
```diff
@@ -0,0 +1,294 @@
...
+        /// <summary>
+        /// PCM_U32BE
+        /// </summary>
+        PcmU32be
+    }
+
+    /// <summary>
+    /// Enumerator for error type from TVESPlayer
+    /// </summary>
+    public enum ErrorType
```
Comment (myroot) [inline src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs:211]: It is too common naming
```suggestion
    public enum ESPlayerErrorType

```

**Review by myroot — CHANGES_REQUESTED** [review]
Please rebase to the latest master branch

File: src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs (hunk 39 above)
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs (hunk 38 above)
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/TVESEnumerations.cs]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/TVESPlayer.cs (hunk 37 above)
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/TVESPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/Renderer/MediaPlayerImpl.cs [hunk 40]
```diff
@@ -41,9 +41,18 @@ public class MediaPlayerImpl : IPlatformMediaPlayer
 
         public MediaPlayerImpl()
         {
-            _player = new Player();
+            if(_player == null)
+                _player = CreateMediaPlayer();
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/Renderer/MediaPlayerImpl.cs:45]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/Renderer/TVESExtensions.cs [hunk 41]
```diff
@@ -0,0 +1,319 @@
...
+        }
+
+        public static TM.ESPacket ToESPacket(this ESPacket packet)
+        {
+            var nPacket = new TM.ESPacket();
+            nPacket.type = packet.StreamType.ToStreamType();
+            nPacket.duration = packet.Duration;
+            nPacket.buffer = packet.Buffer;
+            nPacket.pts = packet.PresentationTimeStamp;
+            return nPacket;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/Renderer/TVESExtensions.cs]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/MediaPlayer.cs [hunk 42]
```diff
@@ -103,7 +103,9 @@ public class MediaPlayer : Element
         /// </summary>
         public MediaPlayer()
         {
-            _impl = DependencyService.Get<IPlatformMediaPlayer>(fetchTarget: DependencyFetchTarget.NewInstance);
+            if(_impl == null)
+                _impl = CreateMediaPlayerImpl();
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/MediaPlayer.cs]: Fixed it. Thank you.

File: src/Tizen.TV.UIControls.Forms/ITVESPlayer.cs [hunk 43]
```diff
@@ -0,0 +1,172 @@
...
+    /// <summary>
+    /// Represents one of es packet which includes handle of data inside the trust zone,
+    /// contains stream type, handle for encrypted es data packet, handle size, pts and duration.
+    /// </summary>
+    public class ESHandlePacket
+    {
+        /// <summary>
+        /// Stream type of ESPacket
+        /// </summary>
+        public StreamType StreamType;
```
[the remark this answers is not in the document]
Comment (yourina) [reply src/Tizen.TV.UIControls.Forms/ITVESPlayer.cs]: Fixed it. Thank you.

**Comment by ppaneksamsung** [comment]
@yourina do you have any samples of using your API?

**Pull request merged**